### PR TITLE
Convert user directory handler and related classes to async/await

### PR DESCRIPTION
**Pull request by clokep**
This converts:
* User directory handler
* Stats handler
* State deltas handler

Note that I did all three since the stats handler sub-classes state deltas handler...although I realized there was no reason for it to be doing that so I removed it. Kind of out of scope for this PR though.

I added a `ensureDeferred` to the register handler to limit the scope of the PR.

### CODE REVIEW

File: synapse/handlers/user_directory.py [hunk 1]
```diff
@@ -306,26 +300,24 @@ def _handle_room_publicity_change(self, room_id, prev_event_id, event_id, typ):
         # being added multiple times. The batching upserts shouldn't make this
         # too bad, though.
         for user_id, profile in iteritems(users_with_profile):
-            yield self._handle_new_user(room_id, user_id, profile)
+            await self._handle_new_user(room_id, user_id, profile)
 
-    @defer.inlineCallbacks
-    def _handle_local_user(self, user_id):
+    async def _handle_local_user(self, user_id):
```
Comment (richvdh) [inline synapse/handlers/user_directory.py:305]: this seems to be unused :-S

File: synapse/handlers/stats.py [hunk 2]
```diff
@@ -135,13 +131,12 @@ def _unsafe_process(self):
...
-    @defer.inlineCallbacks
-    def _handle_deltas(self, deltas):
+    async def _handle_deltas(self, deltas):
         """Called with the state deltas to process
 
         Returns:
-            Deferred[tuple[dict[str, Counter], dict[str, counter]]]
-            Resovles to two dicts, the room deltas and the user deltas,
+            tuple[dict[str, Counter], dict[str, counter]]
+            Resolves to two dicts, the room deltas and the user deltas,
```
Comment (richvdh) [inline synapse/handlers/stats.py:139]: ftr, "Resolves" only really makes sense when we're talking about deferreds
```suggestion
            Two dicts: the room deltas and the user deltas,
```

File: synapse/handlers/user_directory.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/user_directory.py:305]: Ah, yeah I meant to mention that. Does it seem OK to remove it here?

Comment (clokep) [reply synapse/handlers/user_directory.py]: Talked about this in #synapse-dev:matrix.org and we decided it is fine to remove.

**Pull request merged**